### ISSUE TITLE
Absolute beginner's guide to install this?

### ISSUE THREAD

**Issue by DMahalko** [reactiflux/discord-irc#314]
Apparently this software is intended to be used by people who are already experts with node.js. I have never used it before.

I have no idea what I am doing. I am typing commands that are twiddling unknown settings deep in the operating system, and I don't like this at all. I don't want to end up having to format and reinstall because of unknown borkage that might be occurring due to trying to make this work.

Can you please break down the install process for this in as few steps as possible? I am using Ubuntu 16.04.

Steps so far.
sudo apt-get install nodejs
sudo apt-get install npm
sudo npm install -g discord-irc
.... this magically does "something", and it somehow knows what "discord-irc" is and doesn't tell me how it knows what it is. I don't know what's going on here. I might be installing malware. No idea. Fails.

└─┬ discord-irc@2.4.2
  ├── UNMET PEER DEPENDENCY bufferutil@^3.0.0
  ├── UNMET PEER DEPENDENCY erlpack@hammerandchisel/erlpack
  ├── UNMET PEER DEPENDENCY libsodium-wrappers@^0.5.1
  ├── UNMET PEER DEPENDENCY node-opus@^0.2.5
  ├── UNMET PEER DEPENDENCY opusscript@^0.0.3
  ├── UNMET PEER DEPENDENCY sodium@^2.0.1
  └── UNMET PEER DEPENDENCY uws@^0.14.1

Okay. Um. now what. Google. Maybe auto-install can fix up whatever is needed?
sudo npm install -g auto-install
Install succeeded

Try discord-irc again. Nope, still fails on dependencies.

Well let's try those one at a time.......

sudo npm install -g uws@^0.14.1
Succeeds
sudo npm install -g  sodium@^2.0.1
Fails, bunch of error messages.
sudo npm install -g opusscript@^0.0.3
Succeeds
sudo npm install -g node-opus@^0.2.5
Succeeds
sudo npm install -g bufferutil@^3.0.0
Fails.

?? Google search, someone suggests make sure package "build-essential" is installed.
sudo apt-get install build-essential
"This is already installed. Install type changed to manual."
Um... ok...
Still doesn't work. Another google search... apparently Ubuntu doesn't make a file named "node" in /usr/bin.
sudo ln -s /usr/bin/nodejs /usr/bin/node

Try again..
sudo npm install -g  bufferutil
Succeeds

sudo npm install -g node-opus
Succeeds
sudo npm install -g opusscript
Succeeds
sudo npm install -g sodium
Fails.
sudo npm install -g libsodium-wrappers
Succeeds (?????)

Try again.
 sudo npm install -g sodium
Fails.

Is there some ubuntu package I need to use instead or something?
apt-cache search sodium
sudo apt-get install libsodium18
Succeeds.
sudo npm install -g sodium
Fails
Try developer package instead.
sudo apt-get install libsodium-dev
Try again.
sudo npm install -g sodium
Well..... it isn't the same error as before but now I get a whole pile of "make" errors.

Google has no answers to these. I just find pastebins without comments.

Try this whole thing over again...
sudo npm install -g discord-irc

?? I still get a list of dependency errors of things that supposedly installed.

└─┬ discord-irc@2.4.2
  ├── UNMET PEER DEPENDENCY bufferutil@^3.0.0
  ├── UNMET PEER DEPENDENCY erlpack@hammerandchisel/erlpack
  ├── UNMET PEER DEPENDENCY libsodium-wrappers@^0.5.1
  ├── UNMET PEER DEPENDENCY node-opus@^0.2.5
  ├── UNMET PEER DEPENDENCY opusscript@^0.0.3
  ├── UNMET PEER DEPENDENCY sodium@^2.0.1
  └── UNMET PEER DEPENDENCY uws@^0.14.1

Try uws again.
sudo npm install -g uws@^0.14.1
Succeeds.... but.... it says

/usr/local/lib
└── uws@0.14.5

So it actually installed a newer version of uws, but apparently discord-irc is demanding an older version that I don't know how to install. And sodium still is not working either.

Also I've now done tons of crap to Ubuntu with sudo, and with no idea how to back out of all this.

**Comment by Throne3d** [comment]
So, you installed a bunch of things with `apt-get` – `sudo` makes your commands privileged (like running in admin mode), `apt-get` is the operating system package manager.

You've also installed a bunch of things with `npm`, the *Node* package manager (see initials), which fetches packages from `npmjs.com`. Packages are uploaded there, and can be seen there ([discord-irc here](https://www.npmjs.com/package/discord-irc), individual files not clearly visible).

I have no idea what `auto-install` is. I don't think it does anything to `npm` installations. Peer dependencies, or at least these ones (it may not hold for the general case), are optional. Most of these are from the upstream `discord.js` library, which is the large snippet of code that we don't maintain that we rely on to communicate with Discord. I suppose we could, [as they have](https://github.com/hydrabolt/discord.js#installation), put a note in the README that unmet peer dependencies are optional. You did not need to modify anything, from what I can tell, to get it to work past that point.

To back out of this, you could look into the documentation on `npm`. This can be done, relatively unsurprisingly, with `npm help` or `npm --help`. It lists a bunch of commands you might find useful. For example, `uninstall`. The `-g` flag makes it apply globally, instead of to the folder you're currently looking at. (It's called a flag because it modifies the behavior, and is indicated by the initial dash; this is a standard thing with lots of UNIX command-line tools, UNIX including Linux and macOS and certain other derivatives.) You may want to go through and remove the `npm` packages you manually installed, presumably with `sudo npm uninstall -g [package-name]` (remove the square brackets).

You can also remove the various Ubuntu packages, with `apt-get`. I'm not on Ubuntu, so I can't be sure, but it looks like you want `remove` – that is, `sudo apt-get remove [package]` – for the packages you don't need. You don't need, as far as I know, any of the ones you installed past your first attempt to install `discord-irc`, which appears not to have failed, but instead just mentioned unmet peer dependencies.

This software is intended, I believe, to be used by people who have passing familiarity with node.js, seeing as it uses node's package manager. I'm not sure if you're coming from a Windows background, here, but you could do this on Windows and you might feel more comfortable there, as node is multi-platform. I do not believe `npm` tweaks deep internal system settings except insofar as it modifies your path and presumably adds configuration options places that configuration options are usually added.

I'd recommend going through and uninstalling (with `npm` and `apt-get`) the packages you don't need (those that you went through and added), and then try using the command `discord-irc --help` to make sure it responds properly. You should then be able to proceed with the instructions as given.

**Summary**: it looks like you took the unmet peer dependencies as required, when they're actually not; we should probably add this to the README.

**Comment by ekmartin** [comment]
The install might have completed successfully - try running `discord-irc`. If not you probably need to upgrade your Node.js version, as apt's default version is quite old: https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions

**Comment by DMahalko** [comment]
Okay, I found how to delete everything and start over, and to update the node.js. I also managed to make all the warnings go away, except it is still complaining about a missing "package.json" and I don't know what to do next.

================
(Removed Everything, cleaned up /usr/lib  /usr/bin ...)

curl -sL https://deb.nodesource.com/setup_8.x | sudo -E bash -
sudo apt-get install nodejs
... (update to 8.x nodejs)

npm install discord-irc
... (fails, unicode/ucsdet.h: No such file or directory)

sudo apt-get install libicu-dev
npm install discord-irc
. . . (it gets past the unicode error, but still has problems)

npm WARN saveError ENOENT: no such file or directory, open '/home/foobar/discord-irc/package.json'
npm WARN enoent ENOENT: no such file or directory, open '/home/foobar/discord-irc/package.json'
npm WARN discord.js@11.1.0 requires a peer of bufferutil@^3.0.0 but none was installed.
npm WARN discord.js@11.1.0 requires a peer of erlpack@hammerandchisel/erlpack but none was installed.
npm WARN discord.js@11.1.0 requires a peer of node-opus@^0.2.5 but none was installed.
npm WARN discord.js@11.1.0 requires a peer of opusscript@^0.0.3 but none was installed.
npm WARN discord.js@11.1.0 requires a peer of sodium@^2.0.1 but none was installed.
npm WARN discord.js@11.1.0 requires a peer of libsodium-wrappers@^0.5.1 but none was installed.
npm WARN discord.js@11.1.0 requires a peer of uws@^0.14.1 but none was installed.
npm WARN discord-irc No description
npm WARN discord-irc No repository field.
npm WARN discord-irc No README data
npm WARN discord-irc No license field.
=======================

... (Try clearing the warnings, don't know if this helps)

npm install uws
npm install bufferutil
npm install node-opus
npm install opusscript

npm install erlpack
... (? It doesn't like this one)
npm install erlpack@hammerandchisel/erlpack
... (Warning goes away for this one goes away.)

npm install uws@^0.14.1
npm install libsodium-wrappers@^0.5.1
npm install sodium
... (fails, it wants "libtool")

npm install libtool
npm install sodium
... (fails, it doesn't want npm libtool apparently)

sudo apt-get install autotools-dev
npm install sodium
... (fails, doesn't seem to help)

sudo apt-get install autoconf
npm install sodium
... (fails, doesn't seem to help)

sudo apt-get install libtool
npm install sodium
... (FINALLY sodium install succeeds)

=======================
But I still don't know what I need to do next.

~/discord-irc$ npm install discord-irc
npm WARN saveError ENOENT: no such file or directory, open '/home/foobar/discord-irc/package.json'
npm WARN enoent ENOENT: no such file or directory, open '/home/foobar/discord-irc/package.json'
npm WARN discord-irc No description
npm WARN discord-irc No repository field.
npm WARN discord-irc No README data
npm WARN discord-irc No license field.

+ discord-irc@2.4.2
updated 1 package in 1.356s
=========================


If I go to ~/discord-irc/node_modules/discord-irc   I see this:
:~/discord-irc/node_modules/discord-irc$ ls -glot
total 188
-rw-rw-r-- 1   2514 Oct  1 09:12 package.json
drwxrwxr-x 2   4096 Oct  1 09:12 dist
-rw-rw-r-- 1   8130 Aug 31 02:41 CHANGELOG.md
-rw-rw-r-- 1 163776 Aug 31 02:41 package-lock.json
-rw-rw-r-- 1   4400 Jul 18 23:45 README.md
-rw-rw-r-- 1   1094 May  6  2016 LICENSE


So..... I don't know what this "package.json" it is still complaining about is for.
Let's copy package.json  and put it in ~/discord-irc and try again.

~/discord-irc/node_modules/discord-irc$ cp package.json ../..
~/discord-irc/node_modules/discord-irc$ cd ../..
~/discord-irc$ npm install discord-irc
npm ERR! code ENOSELF
npm ERR! Refusing to install package with name "discord-irc" under a package
npm ERR! also called "discord-irc". Did you name your project the same
npm ERR! as the dependency you're installing?
npm ERR!
npm ERR! For more information, see:
npm ERR!     <https://docs.npmjs.com/cli/install#limitations-of-npms-install-algorithm>

???

**Comment by ekmartin** [comment]
You don't have to install the other dependencies. Install node.js, then run _only_ `npm install -g discord-irc` (note the `-g` for global). Then you should be able to run `discord-irc --config config.json`.

**Comment by DMahalko** [comment]
Another problem, npm has wrong permissions access, even when run with sudo.
I am looking into it . . .
https://docs.npmjs.com/getting-started/fixing-npm-permissions

```
sudo npm install -g discord-irc
/usr/bin/discord-irc -> /usr/lib/node_modules/discord-irc/dist/index.js

> iconv@2.2.3 install /usr/lib/node_modules/discord-irc/node_modules/iconv
> node-gyp rebuild

gyp ERR! configure error
gyp ERR! stack Error: EACCES: permission denied, mkdir '/usr/lib/node_modules/discord-irc/node_modules/iconv/build'
gyp ERR! System Linux 4.4.0-91-generic
gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /usr/lib/node_modules/discord-irc/node_modules/iconv
gyp ERR! node -v v8.6.0
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok
```

**Comment by Kovaelin** [comment]
Is there a difference between installing discord.js or node.js? I've been using discord.js only and have been getting it to run successfully.

**Comment by Throne3d** [comment]
Uh. I don't think that's possible, since discord.js is a library that's written _in_ node.js, relying on node.js, and does not replace node.js?

The language used is JavaScript, and this runs here on node.js, and discord.js is a library written in JavaScript to keep all the Node ↔ Discord communication in a standardized block of library code, which is then used by this program to actually do things around Discord.

I can only assume you already had node.js installed, or somehow managed to install it while installing discord.js? (Like it says on [discord.js's readme](https://github.com/hydrabolt/discord.js/tree/11.1.0#installation), "Node.js 6.0.0 or newer is required.")

**Comment by Kovaelin** [comment]
Yeah, I do already have node.js installed on my own computer, but I also tried it on Cloud9, G Cloud and Azure, and I could install discord.js without installing node.js first.

**Comment by Throne3d** [comment]
To install it, you would've had to have used `npm`? And then to run this bot you would need Node, and `npm` relies on Node being installed.

So if you installed discord.js, I am rather certain you had node.js installed?

**Comment by Kovaelin** [comment]
I think you're right then. I guess most cloud services just have node.js ready.

**Comment by DMahalko** [comment]
Followup comment (this can stay closed).

I was able to get this working, though the npm permission thing is weird.  Running "sudo npm install -g discord-js" somehow manages to fail.

"npm" is running as root and it doesn't have write permission? It appears npm is screwed up in some manner if sudo doesn't give it everything it needs.

I have no idea if npm itself is a script or a linux executable in order to try to analyze this permissions problem.

Meanwhile a youtube video about this global install problem, from the npm creators, suggests reassigning a bunch of core system directories in /usr from "root" to your own username, which has to be the dumbest thing I've ever heard. Root owns all that stuff in /usr for a very good reason.

In the end I made a new custom directory for global nodejs stuff named "/usr/nodejs" and set that as the default npm install path. So now npm -g works without sudo, which still seems incredibly dumb for a global install for all users. . . but this is still better than setting /usr/bin to owned by my account. Derp.
